File: 그래프_dfs/leetcode589(Easy)-N-ary Tree Preorder Traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

// DFS
/**
 * @param {Node|null} root
 * @return {number[]}
 */
var preorder = function (root) {
  const ans = []

  const dfs = (node) => {
    if (!node) return
    ans.push(node.val)
    for (const nextNode of node.children) {
      dfs(nextNode)
    }
  }

  dfs(root)

  return ans
}

File: 그래프_dfs/leetcode589(Easy)-N-ary Tree Preorder Traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

# 접근 : DFS
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        ans = []

        def dfs(node):
            if not node:
                return
            ans.append(node.val)
            for next_node in node.children:
                dfs(next_node)

        dfs(root)

        return ans